synapse/handlers/message.py

```diff
@@ -738,8 +746,14 @@ async def create_and_send_nonmember_event(
                     assert event.internal_metadata.stream_ordering
                     return event, event.internal_metadata.stream_ordering
 
+            prev_events = event_dict.get("prev_events")
```

Comment:
Continuing the conversation from https://github.com/matrix-org/synapse/pull/9247#discussion_r573718697:

> This makes me a bit uncomfortable though. I just audited all the calls to `create_and_send_nonmember_event` and they (mostly) seem safe. A questionable call site is the `ModuleApi` -- this gets passed to various pieces of 3rd party code and would potentially give them the same power as an AS. Maybe realistically this is OK, but I'll want opinions of others.

@MadLittleMods responded:
> Should we just add the same `requester.app_service` condition around this to make it more straightforward on what can happen?

I don't think that adding that around this would make it any "safer" since the requester for the Module API is just pulled out of the `event_dict` anyway. I think it is probably OK to leave this though as we treat modules as privileged?

Other thoughts:
* Should we be popping `prev_events` or do we want it to end up in the event dictionary too?
* Please add a comment above this.

Reply:
> Should we be popping `prev_events` or do we want it to end up in the event dictionary too?

No strong opinion. Might be a little unexpected if we have `prev_events` defined on some event_dicts but not others.

> Please add a comment above this.

What kind of comment should be here?

Reply:
I think its safe for module APIs to set `prev_events` (though we shouldn't encourage it), as a) they are trusted bits of code, e.g. if they wanted to they could just wipe the DB quite happily, and b) most of the code needs to deal with weird `prev_events` anyway due to federation